Spotify Controls: Shuffle Play + Play

ron-myers opened the pull request:
## Description

Add new command:

* `Shuffle Play` - launches Spotify, if not open, enables shuffle and hits play if not playing.

Update command:

* `Play` to open Spotify, if not open, before playing.

## Type of change

<!-- Please choose one option and delete others that are not relevant. -->

- Bug fix / improvement for somebody else's extension. Extension author:  @thomaspaulmann

## Screencast

<!-- If you add a new extension or command, include a screencast (or screenshot for a small change). Good screencast will make the review much faster, especially your extension requires registration in other services.  -->


https://user-images.githubusercontent.com/10678906/152661169-8d41ff50-98ac-4a44-9ebe-cb92c01d0798.mov


## Checklist

- [x] I read the [extension guidelines](https://developers.raycast.com/basics/prepare-an-extension-for-store)
- [x] I read the [documentation about publishing](https://developers.raycast.com/basics/publish-an-extension)
- [x] I ran `npm run build` and [tested this distribution build in Raycast](https://developers.raycast.com/basics/prepare-an-extension-for-store#metadata-and-configuration)

thomaspaulmann commented on extensions/spotify-controls/src/play.ts:
```diff
@@ -1,5 +1,23 @@
 
 export default async () => {
-  await runAppleScriptSilently('tell application "Spotify" to play');
+  const script = `
+    tell application "Spotify"
+
+      if not application "Spotify" is running then
+        
+        activate
+
+        repeat until application "Spotify" is running
+          delay 1
+        end repeat
+
+      end if
+
+      play
+        
+    end tell
```
```suggestion
  const script = `
    tell application "Spotify"
      if not application "Spotify" is running then
        activate
        repeat until application "Spotify" is running
          delay 1
        end repeat
      end if
      play
    end tell
```

ron-myers replied:
Good suggestion.  This commit is outdated.

thomaspaulmann commented on extensions/spotify-controls/src/shufflePlay.ts:
```diff
@@ -0,0 +1,32 @@
+import { runAppleScriptSilently } from "./utils";
+
+export default async () => {
+  const script = `
+
+    tell application "Spotify"
+
+      if not application "Spotify" is running then
+        
+        activate
+
+        repeat until application "Spotify" is running
+          delay 1
+        end repeat
+  
+      end if
+
+      set shuffling to true
+
+      set spotify_playing to (player state = playing)
+
+      if not spotify_playing then
+    
+        play
+        
+      end if
+
+    end tell
+  `;
```
```suggestion
  const script = `
    tell application "Spotify"
      if not application "Spotify" is running then
        activate
        repeat until application "Spotify" is running
          delay 1
        end repeat
      end if
      set shuffling to true
      set spotify_playing to (player state = playing)
      if not spotify_playing then
        play
      end if
    end tell
  `;
```

ron-myers replied:
Good suggestion - I adopted that style.

This commit is out of date.

thomaspaulmann reviewed (changes requested):
Thanks for adding the new command and improving the existing "Play" command. A few things to consider:
1. How about adding some safety net to the test if the application started? Maybe just repeat it three times? This makes sure that the script doesn't run endless if the app fails starting. More of an edge case but good to cover it.
2. Would it make sense to start the app for all the commands if it isn't already running? Could also be a preference that is turned on by default. WDYT?

ron-myers commented:
Thanks for the review @thomaspaulmann 

> 1. How about adding some safety net to the test if the application started? Maybe just repeat it three times? This makes sure that the script doesn't run endless if the app fails starting. More of an edge case but good to cover it.

Good thought. I went with 5 seconds(for slower machines) for the max and added a builder to wrap any command in the check to ensure Spotify is open.

See `buildScriptEnsuringSpotifyIsRunning`.  This method may make sense to pull into Raycast's core - I could see it having common use.

> 2. Would it make sense to start the app for all the commands if it isn't already running? Could also be a preference that is turned on by default. WDYT?

Agreed. That seems logical. I added a builder to wrap the command for every command except `current`, which handles when Spotify isn't open. I opted for no setting for now - I couldn't think of why someone would want the new behaviour turned off. Happy to discuss or add a setting if it is seen as necessary.
